Add support for trace inputs to built-in scorers

BenWilson2 opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/BenWilson2/mlflow/pull/17943?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17943/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17943/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/17943/merge
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Adds the ability to pass traces in to built-in scorers. If traces are present, fields are optimistically extracted from traces to populate the required fields for the built-in trace templates. If the fields cannot be extracted, a fallback trace-based template (that utilizes the same base instructions from the scorer's template) that serializes as JSON the Trace object is supplied and called to attempt to produce a directed evaluation. 

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [X] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [X] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [X] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

BenWilson2 committed:
initial

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

dbczumar commented on docs/docs/genai/eval-monitor/scorers/llm-judge/predefined.mdx:
```diff
@@ -68,21 +68,46 @@ results = mlflow.genai.evaluate(
+
+### Automatic Fallback for Complex Traces
+
+When the scorer cannot extract the required fields from a trace (e.g., complex multi-step workflows), it automatically falls back to submitting the serialized trace info to an LLM judge. This ensures evaluation can still proceed even with traces that don't follow the standard input/output pattern.
```
> serialized trace info to an LLM judge

The serialized trace info isn't submitted, right? We use tool calling...

dbczumar commented on mlflow/genai/judges/prompts/correctness.py:
```diff
@@ -30,6 +32,15 @@
 
+CORRECTNESS_TRACE_FALLBACK = f"""{CORRECTNESS_BASE_INSTRUCTIONS}
+
+Extract the question, expected response/facts, and actual response from the following trace and \
```
@BenWilson2 We're using inconsistent terminology. Can we make the terminology consistent with the `CORRECTNESS_BASE_INSTRUCTIONS` that we're appending to?

The base instructions refer to `question`, `claim`, and `document`, but this fallback refers to `question`, `expected response/facts`, and `actual response`.

dbczumar commented on mlflow/genai/judges/prompts/correctness.py:
```diff
@@ -30,6 +32,15 @@
 
 If the claim is fully supported by the document in the context of the question, you must say "The response is correct" in the rationale. If the claim is not fully supported by the document in the context of the question, you must say "The response is not correct"."""  # noqa: E501
 
+CORRECTNESS_TRACE_FALLBACK = f"""{CORRECTNESS_BASE_INSTRUCTIONS}
```
@BenWilson2 Can we incorporate the correctness prompt suffix into the trace fallback as well?

dbczumar replied:
```suggestion
CORRECTNESS_TRACE_FALLBACK_INSTRUCTIONS = f"""{CORRECTNESS_BASE_INSTRUCTIONS}
```

Nit - appending _INSTRUCTIONS to the name is consistent with the approach for the non-fallback prompt, and it makes it clearer that this isn't the full fallback prompt (because `CORRECTNESS_PROMPT_OUTPUT` is appended later)

dbczumar commented on mlflow/genai/judges/prompts/guidelines.py:
```diff
@@ -25,6 +26,21 @@
 
+# Trace-based fallback template when extraction fails
+GUIDELINES_TRACE_FALLBACK_TEMPLATE = (
+    GUIDELINES_BASE_INSTRUCTIONS
+    + """
```
Nit: the formatting here looks a bit odd...should we use f-strings?

dbczumar commented on mlflow/genai/judges/prompts/safety.py:
```diff
@@ -3,19 +3,43 @@
 
-<text>{{content}}</text>
-"""  # noqa: E501
+SAFETY_PROMPT_INSTRUCTIONS = SAFETY_BASE_INSTRUCTIONS
```
Is this line necessary?

dbczumar commented on mlflow/genai/judges/prompts/safety.py:
```diff
@@ -3,19 +3,43 @@
+"""
+
+# Trace-based fallback template when extraction fails
+SAFETY_TRACE_FALLBACK = f"""{SAFETY_BASE_INSTRUCTIONS}
```
```suggestion
SAFETY_TRACE_FALLBACK_INSTRUCTIONS = f"""{SAFETY_BASE_INSTRUCTIONS}
```

dbczumar commented on mlflow/genai/judges/prompts/guidelines.py:
```diff
@@ -25,6 +26,21 @@
 GUIDELINES_PROMPT = GUIDELINES_PROMPT_INSTRUCTIONS + GUIDELINES_PROMPT_OUTPUT
 
+# Trace-based fallback template when extraction fails
+GUIDELINES_TRACE_FALLBACK_TEMPLATE = (
```
```suggestion
GUIDELINES_TRACE_FALLBACK_INSTRUCTIONS = (
```

dbczumar commented on mlflow/genai/scorers/builtin_scorers.py:
```diff
@@ -556,20 +577,52 @@ def get_input_fields(self) -> list[JudgeField]:
+
+            if inputs is None or outputs is None:
+                prompt = get_guidelines_trace_fallback_prompt(self.guidelines)
+                trace_json = trace.to_json(pretty=True)
+                prompt = format_prompt(prompt, trace=trace_json)
```
@BenWilson2 We shouldn't be formatting the trace into the prompt directly, right? We should be using tool calling to read the trace dynamically.

Formatting the full trace into the prompt will very likely blow up the context window in some cases or make the prompt very difficult for the LLM to parse.

Instead, we should do:

```
make_judge(
    name=self.name,
    instructions=prompt,
    model=self.model,
)(trace=trace)
```

BenWilson2 committed:
feedback

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'master' into trace-for-builtin

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 commented:
@dbczumar shifted everything over to the agent-based fallback and addressed comments. I ran through a verification with a capable OpenAI model with manually 'broken' traces to trigger the fallback on a somewhat simple trace (just a handful of spans) and it seemed fairly reasonable for a response.

BenWilson2 committed:
cleanup exception spam

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'master' into trace-for-builtin

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 commented on mlflow/genai/judges/prompts/correctness.py:
```diff
@@ -3,15 +3,17 @@
-<question>{{input}}</question>
-<claim>{{ground_truth}}</claim>
-<document>{{input}} - {{output}}</document>\
+<question>{{{{input}}}}</question>
```
Because they get escaped with fstring parsing, which was introduced to DRY the base prompts.

BenWilson2 commented on mlflow/genai/scorers/builtin_scorers.py:
```diff
@@ -577,20 +652,46 @@ def get_input_fields(self) -> list[JudgeField]:
+
+            if inputs is None or outputs is None:
+                prompt = get_guidelines_trace_fallback_prompt(self.guidelines)
+                judge = make_judge(
```
This would probably make less-powerful LLMs (read: cheaper) usable as the a more complex instruction set for some of the built-in scorers would not be required. It definitely makes sense as a compromise.

BenWilson2 committed:
refactor and simplify

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

dbczumar commented on mlflow/genai/scorers/builtin_scorers.py:
```diff
@@ -36,6 +41,88 @@
+    expectations: dict[str, Any] | None = None
+
+
+def _extract_fields_from_trace(
```
Nit - should this just be called `resolve_scorer_fields` or something? It doesn't always extract the fields from a trace.

dbczumar commented on mlflow/genai/scorers/builtin_scorers.py:
```diff
@@ -36,6 +41,88 @@
+    if inputs is None or outputs is None:
+        extraction_judge = make_judge(
+            name="field_extractor",
+            instructions=(
+                "Extract the following information from {{ trace }}:\n"
+                "- inputs: The user's request or question\n"
+                "- outputs: The system's response\n"
+                'Return as JSON: {"inputs": "...", "outputs": "..."}'
+            ),
```
I'm surprised this works, since the instructions judge base prompt tells the judge to respond with two fields: `value` and `rationale`.

Also, this won't work with models that support structured outputs, since we also enforce `value` and `rationale` output fields via structured outputs.

@BenWilson2 Instead of using `make_judge()` for field extraction, I think we should just:

1. Add an `output_format` field to `invoke_judge_model` that accepts a pydantic model (default is a model with `value & rationale`)
2. Call `invoke_judge_model` with a message history that tells the LLM to extract inputs / outputs and an `output_format` that contains `inputs` and `outputs` fields.

BenWilson2 replied:
Will do - keep it as simple as possible

BenWilson2 committed:
simplify

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
move to local import to fix skinny

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

dbczumar commented on mlflow/genai/judges/utils.py:
````diff
@@ -173,6 +174,37 @@ def add_output_format_instructions(prompt: str, output_fields: list["JudgeField"
     return prompt + output_format_instructions
 
 
+def _strip_markdown_code_blocks(response: str) -> str:
+    """
+    Strip markdown code blocks from LLM responses.
+
+    Some legacy models wrap JSON responses in markdown code blocks (```json...```).
+    This function removes those wrappers to extract the raw JSON content.
+
+    Args:
+        response: The raw response from the LLM
+
+    Returns:
+        The response with markdown code blocks removed
+    """
+    cleaned = response.strip()
+    if not cleaned.startswith("```"):
+        return cleaned
+
+    lines = cleaned.split("\n")
+    start_idx = 0
+    end_idx = len(lines)
+
+    for i, line in enumerate(lines):
+        if i == 0 and line.startswith("```"):
+            start_idx = 1
+        elif line.strip() == "```" and i > 0:
+            end_idx = i
+            break
+
+    return "\n".join(lines[start_idx:end_idx])
````
Is there an existing lib / util that we can use for this or copy from? Worried about  bespoke utilities

dbczumar commented on mlflow/genai/judges/utils.py:
```diff
@@ -387,6 +419,7 @@ def invoke_judge_model(
     assessment_name: str,
     trace: Trace | None = None,
     num_retries: int = 10,
+    output_format: type["BaseModel"] | None = None,
```
Nit - should we use `"pydantic.baseModel"` in the type hint for clarity?

dbczumar commented on mlflow/genai/judges/utils.py:
```diff
@@ -404,6 +437,9 @@ def invoke_judge_model(
         assessment_name: The name of the assessment.
         trace: Optional trace object for context.
         num_retries: Number of retries on transient failures when using litellm.
+        output_format: Optional Pydantic model class for structured output format.
+                       If not provided, defaults to the standard Judge output format
+                       with 'result' and 'rationale' fields.
 
```
@BenWilson2 Can we move this to `_invoke_litellm` instead? Judges should always return `value` / `rationale`. It's also a bit strange if the request / response field extractor returns a `Feedback` instance.

BenWilson2 committed:
move to standalone util

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'master' into trace-for-builtin

Signed-off-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>

dbczumar commented on mlflow/genai/judges/utils.py:
```diff
@@ -511,6 +553,95 @@ def invoke_judge_model(
     return feedback
 
 
+def extract_structured_output(
+    model_uri: str,
+    messages: list["ChatMessage"],
+    output_schema: type[pydantic.BaseModel],
+    trace: Trace | None = None,
+    num_retries: int = 10,
+) -> pydantic.BaseModel:
+    """
+    Extract structured output from an LLM using a Pydantic schema.
```
If I understand correctly, we can also replace the call to `_invoke_litellm` in `invoke_judge_model` with a call to `get_chat_completions_with_structured_output` as well?

dbczumar commented on mlflow/genai/scorers/builtin_scorers.py:
```diff
@@ -26,14 +35,111 @@
+            )
+            inputs = inputs or extracted.inputs
+            outputs = outputs or extracted.outputs
+        except Exception:
+            pass
```
should we log a warning in this case?

dbczumar commented on mlflow/genai/scorers/builtin_scorers.py:
```diff
@@ -694,37 +808,53 @@ def validate_columns(self, columns: set[str]) -> None:
         Evaluate adherence to specified guidelines.
 
+        This scorer can be used in two ways:
+        1. **With a trace**: Pass an MLflow trace object to automatically extract
+           and evaluate inputs, outputs, and expectations from the trace's execution.
+        2. **With explicit parameters**: Directly provide the inputs, outputs, and expectations
+           to evaluate. This approach is more stable and predictable than trace extraction.
```
```suggestion
        1. Pass an MLflow trace object to automatically extract
           and evaluate inputs, outputs, and expectations from the trace's execution.
        2. Directly provide the inputs, outputs, and expectations
           to evaluate. This approach is more stable and predictable than trace extraction.
```

Nit - `explicit parameters` wasn't super clear / intuitive to me. Can we remove this terminology from the PR?

BenWilson2 committed:
feedback

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

B-Step62 commented on mlflow/genai/utils/trace_utils.py:
```diff
@@ -92,6 +92,71 @@ def extract_outputs_from_trace(trace: Trace) -> Any:
 
+def resolve_inputs_from_trace(inputs: Any | None, trace: Trace) -> Any | None:
+    """
+    Extract inputs from trace if not provided.
```
It is a bit strange that we have both `resolve_` and `extract_` and the former actually has docstring "Extract ...". Looking at where `extract_...` is already used, it is indeed inside the very similar method `_resolve_inputs_from_trace`. https://github.com/mlflow/mlflow/blob/e7ab7951ac5b83af003756c80ffb7f84d9a91805/mlflow/genai/judges/instructions_judge/__init__.py#L239

Is there any way we can avoid this duplication?

B-Step62 commented on mlflow/genai/utils/trace_utils.py:
```diff
@@ -184,28 +248,22 @@ def is_none_or_nan(value: Any) -> bool:
 def parse_inputs_to_str(value: Any) -> str:
     """Parse the inputs to a string compatible with the judges API"""
     if is_none_or_nan(value):
-        # The DBX managed backend doesn't allow empty inputs. This is
```
Same here

B-Step62 commented on mlflow/genai/utils/trace_utils.py:
```diff
@@ -184,28 +248,22 @@ def is_none_or_nan(value: Any) -> bool:
 
     NB: This function does not handle pandas.NA.
     """
-    # isinstance(value, float) check is needed to ensure that math.isnan is not called on an array.
```
This is important context, can we keep the comment?

B-Step62 commented on mlflow/genai/utils/trace_utils.py:
```diff
@@ -262,7 +317,6 @@ def _to_dict(obj: Any) -> dict[str, Any]:
     if isinstance(obj, BaseModel):
         return obj.model_dump()
 
-    # Convert to JSON string and then back to dictionary to handle nested objects
```
I think this comment is also useful.

B-Step62 commented on mlflow/genai/utils/trace_utils.py:
```diff
@@ -218,13 +276,10 @@ def parse_outputs_to_str(value: Any) -> str:
     if isinstance(value, str):
         return value
 
-    # PyFuncModel.predict wraps the output in a list
```
Same here, this logic is hard to reason about without knowing how pyfunc works.

B-Step62 commented on mlflow/genai/utils/trace_utils.py:
```diff
@@ -311,28 +364,23 @@ def _get_top_level_retrieval_spans(trace: Trace) -> list[Span]:
     Span C and Span F are NOT top-level because they are children of other retrieval spans.
```
Can we keep comments in this function? They are not llm generated, but is helpful for reading code.

BenWilson2 commented on mlflow/genai/judges/utils.py:
```diff
@@ -1094,13 +1223,27 @@ def _create_litellm_tool_response_message(
 
 
-def _get_judge_response_format() -> dict[str, Any]:
+def _get_judge_response_format(pydantic_format: type["BaseModel"] | None = None) -> dict[str, Any]:
```
This is a fantastic find - this greatly simplifies this and prevents a bunch of complex and fragile parsing logic. 
Pydantic in -> Pydantic out. 

We can just directly construct a model from the field definitions.

BenWilson2 committed:
fixes / optimizations

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
add types

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

B-Step62 commented on mlflow/genai/judges/utils.py:
```diff
@@ -879,23 +996,118 @@ def __exit__(
+
+
 @_suppress_litellm_nonfatal_errors
 def _invoke_litellm(
```
nit: Can we rename it to `invoke_litellm_and_handle_tools` and make `_make_litellm_completion_request` into `invoke_litellim`? To be explicit about the side effect.

B-Step62 reviewed (approved):
LGTM once the comments are addressed

BenWilson2 committed:
trace-for-builtin

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'master' into trace-for-builtin

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>
